File: tests/machine/x/fs/group_by.fs

type Anon1 = {
    name: string
    age: int
    city: string
}
type Anon2 = {
    city: obj
    count: obj
    avg_age: obj
}
let people = [{ name = "Alice"; age = 30; city = "Paris" }; { name = "Bob"; age = 15; city = "Hanoi" }; { name = "Charlie"; age = 65; city = "Paris" }; { name = "Diana"; age = 45; city = "Hanoi" }; { name = "Eve"; age = 70; city = "Paris" }; { name = "Frank"; age = 22; city = "Hanoi" }]
let stats = [ for gKey, gItems in [ for person in people do yield person ] |> List.groupBy (fun person -> person.city) do
    let g = {| key = gKey; items = gItems |}
    yield { city = g.key; count = List.length g.items; avg_age = (List.sum [ for p in g do yield p.age ] / List.length [ for p in g do yield p.age ]) } ]
printfn "%s" "--- People grouped by city ---"
for s in stats do
    printfn "%s" (String.concat " " [string s.city; string ": count ="; string s.count; string ", avg_age ="; string s.avg_age])

File: tests/transpiler/x/py/group_by.py
from dataclasses import dataclass

@dataclass
class Person:
    name: str
    age: int
    city: str

people = [Person("Alice", 30, "Paris"), Person("Bob", 15, "Hanoi"), Person("Charlie", 65, "Paris"), Person("Diana", 45, "Hanoi"), Person("Eve", 70, "Paris"), Person("Frank", 22, "Hanoi")]
@dataclass
class Stat:
    city: str
    count: int
    avg_age: float

@dataclass
class StatsGroup:
    key: str
    items: list

_stats_groups = {}
for person in people:
    g = _stats_groups.setdefault(person.city, StatsGroup(person.city, []))
    g.items.append(person)
stats = []
for g in _stats_groups.values():
    stats.append(Stat(city=g.key, count=len(g.items), avg_age=sum(p.age for p in g.items) / len(g.items)))
print("--- People grouped by city ---")
for s in stats:
    print(f"{s.city}: count = {s.count}, avg_age = {s.avg_age}")
